ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java

```diff
@@ -191,22 +193,22 @@ protected void onAccountUpdateNotification(Collection<Account> updatedAccounts)
   /**
    * Parse the json config for {@link QuotaEnforcer} and {@link QuotaSource} factory pair and return them in a {@link Map}.
    * @param quotaEnforcerSourceJson json config string.
-   * @return Map of {@link QuotaEnforcer} and {@link QuotaSource} factory pair.
+   * @return List of {@link QuotaEnforcer} and {@link QuotaSource} factory {@link Pair}s.
    */
-  private Map<String, String> parseQuotaEnforcerAndSourceInfo(String quotaEnforcerSourceJson) {
+  private List<Pair<String, String>> parseQuotaEnforcerAndSourceInfo(String quotaEnforcerSourceJson) {
     if (quotaEnforcerSourceJson.isEmpty()) {
-      return Collections.emptyMap();
+      return Collections.emptyList();
     }
-    Map<String, String> quotaEnforcerSourceMap = new HashMap<>();
+    List<Pair<String, String>> quotaEnforcerSourcePairs = new ArrayList<>();
     JSONObject root = new JSONObject(quotaEnforcerSourceJson);
     JSONArray all = root.getJSONArray(QuotaConfig.QUOTA_ENFORCER_SOURCE_PAIR_INFO_STR);
     for (int i = 0; i < all.length(); i++) {
       JSONObject entry = all.getJSONObject(i);
       String enforcer = entry.getString(QuotaConfig.ENFORCER_STR);
       String source = entry.getString(QuotaConfig.SOURCE_STR);
-      quotaEnforcerSourceMap.put(enforcer, source);
+      quotaEnforcerSourcePairs.add(new Pair<>(enforcer, source));
```

Comment:
May I know in which case the QuotaEnforcerFactory will mapping to multiple QuotaSourceFactory? Is this for migration? Also, can you update the test corresponding to the changes?

Reply:
Have same question as Sophie. When would we have multiple instances of same enforcer type?

Reply:
The methods changes are internal private methods of this class. All the tests that rely on AmbryQuotaManager, will automatically test the new changes. More explicit tests that exercise the logic for multiple QuotaEnforcer instances of the same type are present in the PR#2052.

As for the question about when multiple instances of the same enforcer type might be used, I realized this when writing some tests. It's theoretically possible that a same QuotaEnforcer implementation might differ in functionality based on some arguments passed to it while construction. In that case there will be two different instances but with the same type. In PR#2052, `AmbryQuotaManagerTest::buildAmbryQuotaManager` and `AmbryQuotaManagerTest::buildTestQuotaEnforcerSourceInfoPairJson` has an example for such a scenario.